### DISCUSSION TITLE
2.0 ORM Automap delete cascade does not work

### DISCUSSION THREAD

**Discussion by emburse-manikamidha** [sqlalchemy/sqlalchemy#11918] [Usage Questions]
Hello everyone, please read below for my use case in MySQL : 

```
class Parent(Base):
    __tablename__ = 'parent'

    ParentID = Column(Integer, primary_key=True)
    children = relationship("Child", back_populates="parent2",
                            cascade="all, delete", passive_deletes=True)


class Child(Base):
    __tablename__ = 'child'

    ChildID = Column(Integer, primary_key=True)
    SchoolID = Column(Integer, primary_key=True)
    ParentID = Column(Integer, ForeignKey('Parent.ParentID', ondelete='CASCADE'), primary_key=True,
                       nullable=False)
    parent2 = relationship("Parent", back_populates="children")
```

Above is my structure. I have created models using Automap and had to declare relationships as shown above only because I was getting foreign key constraint error.

```
(pymysql.err.IntegrityError) (1451, 'Cannot delete or update a parent row: a foreign key constraint fails (`database`.`Child`, CONSTRAINT `constraint_name` FOREIGN KEY (`ParentID`) REFERENCES `Parent` (`ParentID`))')
[SQL: DELETE FROM `parent` WHERE `parent`.`ParentID` = %(ParentID)s]
```

FYI, the tables created in the work database do not have 'on delete cascade'. My understanding is that passive_deletes=True will take care of this.

This is how I am deleting : 
```
model_names = [Parent]. # just using Parent as expecting the child rows to be automatically deleted.
color = "red"

for model_name in model_names:
    statement = select(model_name).filter_by(ColorID=color)
    rows = session.execute(statement).all()
    total_rows_list += rows

for row_tuple in total_rows_list:
    rows_to_delete.append(row_tuple[0])

for row in rows_to_delete:
     session.delete(row)

session.commit()
```

Any idea why I am getting this error ?

**Comment by emburse-manikamidha** [comment]
Even if I do this : 
```
class Parent(Base):
    __tablename__ = 'parent'

    ParentID = Column(Integer, primary_key=True)


class Child(Base):
    __tablename__ = 'child'

    ChildID = Column(Integer, primary_key=True)
    SchoolID = Column(Integer, primary_key=True)
    ParentID = Column(Integer, primary_key=True)

model_names = [Child, Parent]
color = "red"

for model_name in model_names:
    statement = select(model_name).filter_by(ColorID=color)
    rows = session.execute(statement).all()
    total_rows_list += rows

for row_tuple in total_rows_list:
    rows_to_delete.append(row_tuple[0])

for row in rows_to_delete:
     session.delete(row)

session.commit()
```

I get the same error -
 ```
(pymysql.err.IntegrityError) (1451, 'Cannot delete or update a parent row: a foreign key constraint fails (`database`.`Child`, CONSTRAINT `constraint_name` FOREIGN KEY (`ParentID`) REFERENCES `Parent` (`ParentID`))')
[SQL: DELETE FROM `parent` WHERE `parent`.`ParentID` = %(ParentID)s]
```

Why am I getting this error even though I am deleting child rows before parent rows ?

**Comment by zzzeek** [comment]
> Above is my structure. I have created models using Automap 

this means you are using table reflection, your tables already exist.  

> FYI, the tables created in the work database do not have 'on delete cascade'. 

right, they dont have the directive, so there's no database-side cascading of deletes

> My understanding is that passive_deletes=True will take care of this.

no, that is the opposite of what that means.  passive_deletes would need to remain at its default of `False` here, unless you can use ALTER TABLE to add `ON DELETE CASCADE` in your database.   Additionally, SQLAlchemy's "in-python cascading" feature requires the use of a `relationship()`, if you use models that have no relationships then SQLAlchemy has no way to delete particular rows before others.

**Reply by emburse-manikamidha** [reply]
@zzzeek Thank you.  I did not understand - 'Additionally, SQLAlchemy's "in-python cascading" feature requires the use of a relationship(), if you use models that have no relationships then SQLAlchemy has no way to delete particular rows before others.'.

My tables have foreign key relationships. Automap detects them automatically.

**Reply by zzzeek** [reply]
sure, your "even if I do this" second example did not show automap, it showed explicit mappings

**Reply by emburse-manikamidha** [reply]
Thanks @zzzeek I think I just made it work. Just one thing I wanted to double check, if I have on delete = cascade, I should just pass the parent table/ class to delete, right ? Eg:

```
class Parent(Base):
    __tablename__ = 'tbl_Message'

    ParentID = Column(Integer, primary_key=True)


class Child(Base):
    __tablename__ = 'tbl_MessageDescription'

    ChildID = Column(Integer, primary_key=True)
    LanguageID = Column(Integer, primary_key=True)
    ParentID = Column(Integer, ForeignKey('Parent.MessageID', ondelete='CASCADE'), primary_key=True,
                       nullable=False)

model_name = [Parent]
for name in model_name:
# delete child and parent rows 
```

**Reply by zzzeek** [reply]
if your database schema also has the "ON DELETE CASCADE" set up for those tables, then the DB will handle deleting rows from Child whenever a Parent row is deleted, so from the client perspective you can emit DELETE for just the parent, or for the two tables in any order.
